File: bin/analyze_primer_alignments.R
#!/usr/bin/env Rscript

# Analyze primer alignment results from bowtie2
suppressPackageStartupMessages({
  library(optparse)
  library(Rsamtools)
  library(GenomicAlignments)
  library(Biostrings)
})

opt_list <- list(
  make_option("--alignment_bam", type="character", help="Input BAM file from bowtie2 alignment"),
  make_option("--primers_tsv", type="character", help="Original primers TSV file"),
  make_option("--transcriptome_fasta", type="character", default=NULL, help="Transcriptome FASTA file for gene name mapping"),
  make_option("--transcript_mapping", type="character", default=NULL, help="Pre-built transcript-to-gene mapping TSV file (faster alternative to FASTA)"),
  make_option("--out_report", type="character", default="primer_alignment_report.tsv", help="Output alignment report"),
  make_option("--out_summary", type="character", default="primer_alignment_summary.tsv", help="Detailed alignment summary per primer")
)

opt <- parse_args(OptionParser(option_list=opt_list))

if (is.null(opt$alignment_bam) || is.null(opt$primers_tsv)) {
  stop("Both --alignment_bam and --primers_tsv are required")
}

# Function to parse gene names from transcriptome FASTA headers
parse_gene_names <- function(transcriptome_fasta = NULL, transcript_mapping = NULL) {
  
  if (!is.null(transcript_mapping) && file.exists(transcript_mapping)) {
    # Use pre-built mapping file (much faster)
    cat("Loading pre-built transcript-to-gene mapping...\n")
    mapping_data <- read.table(transcript_mapping, sep="\t", header=TRUE, stringsAsFactors=FALSE, quote="")
    
    if (!"transcript_id" %in% colnames(mapping_data) || !"gene_name" %in% colnames(mapping_data)) {
      stop("Mapping file must contain 'transcript_id' and 'gene_name' columns")
    }
    
    # Convert to list format for compatibility
    transcript_to_gene <- as.list(setNames(mapping_data$gene_name, mapping_data$transcript_id))
    cat(sprintf("Loaded mapping for %d transcripts\n", length(transcript_to_gene)))
    return(transcript_to_gene)
  }
  
  # Fallback to FASTA parsing if no mapping file
  if (is.null(transcriptome_fasta) || !file.exists(transcriptome_fasta)) {
    cat("Warning: Transcriptome FASTA not provided or not found. Gene names will not be available.\n")
    return(NULL)
  }
  
  cat("Parsing gene names from transcriptome FASTA (this may take time)...\n")
  tryCatch({
    # Read FASTA sequences
    fasta_seqs <- readDNAStringSet(transcriptome_fasta)
    transcript_to_gene <- list()
    
    # Parse each header to extract gene name
    for (i in seq_along(fasta_seqs)) {
      full_header <- names(fasta_seqs)[i]
      
      # Extract transcript ID (first part before space)
      transcript_id <- strsplit(full_header, " ")[[1]][1]
      
      # Extract gene name from header (assuming format like "ENST00000456328 gene=DDX11L2")
      gene_name <- "Unknown"
      if (grepl("gene=", full_header)) {
        # Use more robust regex to extract gene name
        gene_match <- regmatches(full_header, regexpr("gene=([A-Za-z0-9_.-]+)", full_header))
        if (length(gene_match) > 0) {
          gene_name <- sub("gene=", "", gene_match)
        }
      }
      
      # Store mapping using clean transcript ID
      transcript_to_gene[[transcript_id]] <- gene_name
      
      # Progress indicator for large files
      if (i %% 10000 == 0) {
        cat(sprintf("Processed %d/%d sequences\n", i, length(fasta_seqs)))
      }
    }
    
    cat("Parsed", length(transcript_to_gene), "transcript-to-gene mappings\n")
    
    # Debug: check if a few specific transcripts are in the mapping
    test_transcripts <- c("ENST00000521262", "ENST00000524349", "ENST00000371437")
    found_count <- 0
    for (test_id in test_transcripts) {
      if (test_id %in% names(transcript_to_gene)) {
        found_count <- found_count + 1
        if (found_count <= 2) {
          cat("  Example mapping:", test_id, "->", transcript_to_gene[[test_id]], "\n")
        }
      }
    }
    
    return(transcript_to_gene)
    
  }, error = function(e) {
    cat("Error parsing transcriptome FASTA:", e$message, "\n")
    return(NULL)
  })
}

# Read original primer data
primers <- read.delim(opt$primers_tsv, stringsAsFactors = FALSE)
primers$primer_id <- paste0(
  primers$gene_id, "|",
  "idx", primers$primer_index, "|",
  primers$primer_type, "|",
  "strand", primers$gene_strand
)

# Parse transcriptome FASTA for gene name mapping
transcript_to_gene <- parse_gene_names(opt$transcriptome_fasta, opt$transcript_mapping)

# Read alignment results
bam_file <- BamFile(opt$alignment_bam)

# Try to read alignments with query names
tryCatch({
  # Use scanBam to get alignments (exclude MAPQ since it's meaningless with -a flag)
  param <- ScanBamParam(what = c("qname", "rname", "pos", "qwidth", "strand"))
  bam_data <- scanBam(bam_file, param = param)[[1]]
  
  cat("Read", length(bam_data$qname), "alignments from BAM file\n")
  
  # Handle case where there are no alignments
  if (length(bam_data$qname) == 0) {
    cat("Warning: No alignments found in BAM file\n")
    alignment_df <- data.frame(
      primer_id = character(0),
      target = character(0),
      start = integer(0),
      end = integer(0),
      width = integer(0),
      strand = character(0),
      stringsAsFactors = FALSE
    )
  } else {
    # Create alignment dataframe from scanBam results
    alignment_df <- data.frame(
      primer_id = bam_data$qname,
      target = as.character(bam_data$rname),
      start = bam_data$pos,
      end = bam_data$pos + bam_data$qwidth - 1,
      width = bam_data$qwidth,
      strand = as.character(bam_data$strand),
      stringsAsFactors = FALSE
    )
    
    # Remove any rows with NA values
    alignment_df <- alignment_df[!is.na(alignment_df$primer_id), ]
  }
  
}, error = function(e) {
  cat("Error reading BAM file with scanBam, trying readGAlignments...\n")
  
  # Fallback to original method
  alignments <- readGAlignments(bam_file)
  cat("Read", length(alignments), "alignments from BAM file\n")
  
  # Handle case where there are no alignments
  if (length(alignments) == 0) {
    cat("Warning: No alignments found in BAM file\n")
    alignment_df <<- data.frame(
      primer_id = character(0),
      target = character(0),
      start = integer(0),
      end = integer(0),
      width = integer(0),
      strand = character(0),
      stringsAsFactors = FALSE
    )
  } else {
    # Extract alignment statistics
    primer_names <- names(alignments)
    if (is.null(primer_names)) {
      primer_names <- paste0("read_", seq_along(alignments))
    }
    
    alignment_df <<- data.frame(
      primer_id = primer_names,
      target = as.character(seqnames(alignments)),
      start = start(alignments),
      end = end(alignments),
      width = width(alignments),
      strand = as.character(strand(alignments)),
      stringsAsFactors = FALSE
    )
  }
})

# Count alignments per primer
alignment_counts <- table(alignment_df$primer_id)

# Create comprehensive report
report <- data.frame(
  gene_id = primers$gene_id,
  primer_index = primers$primer_index,
  primer_type = primers$primer_type,
  primer_sequence = primers$primer_sequence,
  gene_strand = primers$gene_strand,
  primer_id = primers$primer_id,
  num_alignments = as.numeric(alignment_counts[primers$primer_id]),
  stringsAsFactors = FALSE
)

# Replace NA with 0 for primers with no alignments
report$num_alignments[is.na(report$num_alignments)] <- 0

# Write report (no alignment quality or MAPQ columns)
write.table(report, opt$out_report, sep = "\t", quote = FALSE, row.names = FALSE)

# Create detailed alignment summary (gene, primer, strand, transcript alignments)
if (nrow(alignment_df) > 0) {
  # Use all alignments since MAPQ is meaningless with bowtie2 -a
  cat("Processing", nrow(alignment_df), "total alignments\n")
  
  # Merge alignment data with primer information
  detailed_summary <- merge(alignment_df, primers[, c("primer_id", "gene_id", "primer_index", 
                                                       "primer_type", "primer_sequence", "gene_strand")], 
                           by = "primer_id", all.x = TRUE)
  
  if (nrow(detailed_summary) > 0) {
    # Add gene names from transcriptome mapping
    if (!is.null(transcript_to_gene)) {
      detailed_summary$aligned_gene_name <- sapply(detailed_summary$target, function(transcript_id) {
        gene_name <- transcript_to_gene[[transcript_id]]
        if (is.null(gene_name)) gene_name <- "Unknown"
        return(gene_name)
      })
    } else {
      detailed_summary$aligned_gene_name <- "Unknown"
    }
    
    # Reorder columns for better readability (remove MAPQ)
    detailed_summary <- detailed_summary[, c("gene_id", "primer_index", "primer_type", 
                                            "primer_sequence", "gene_strand", "target", 
                                            "aligned_gene_name", "start", "end", "width", 
                                            "strand")]
    
    # Sort by gene, primer index, and number of mismatches (best alignments first)
    detailed_summary <- detailed_summary[order(detailed_summary$gene_id, 
                                              detailed_summary$primer_index), ]
    
    # Rename columns for clarity
    names(detailed_summary)[names(detailed_summary) == "target"] <- "aligned_transcript"
    names(detailed_summary)[names(detailed_summary) == "start"] <- "alignment_start"
    names(detailed_summary)[names(detailed_summary) == "end"] <- "alignment_end"
    names(detailed_summary)[names(detailed_summary) == "width"] <- "alignment_length"
    names(detailed_summary)[names(detailed_summary) == "strand"] <- "alignment_strand"
    
    # Write detailed summary
    write.table(detailed_summary, opt$out_summary, sep = "\t", quote = FALSE, row.names = FALSE)
    
    cat("Detailed alignment summary written to:", opt$out_summary, "\n")
  } else {
    # Create empty detailed summary file  
    empty_summary <- data.frame(
      gene_id = character(0),
      primer_index = integer(0),
      primer_type = character(0),
      primer_sequence = character(0),
      gene_strand = character(0),
      aligned_transcript = character(0),
      aligned_gene_name = character(0),
      alignment_start = integer(0),
      alignment_end = integer(0),
      alignment_length = integer(0),
      alignment_strand = character(0)
    )
    write.table(empty_summary, opt$out_summary, sep = "\t", quote = FALSE, row.names = FALSE)
    
    cat("No high-quality alignments found - empty detailed summary written to:", opt$out_summary, "\n")
  }
} else {
  # Create empty detailed summary file
  empty_summary <- data.frame(
    gene_id = character(0),
    primer_index = integer(0),
    primer_type = character(0),
    primer_sequence = character(0),
    gene_strand = character(0),
    aligned_transcript = character(0),
    aligned_gene_name = character(0),
    alignment_start = integer(0),
    alignment_end = integer(0),
    alignment_length = integer(0),
    alignment_strand = character(0)
  )
  write.table(empty_summary, opt$out_summary, sep = "\t", quote = FALSE, row.names = FALSE)
  
  cat("No alignments found - empty detailed summary written to:", opt$out_summary, "\n")
}

cat("Alignment analysis complete\n")
cat("Report written to:", opt$out_report, "\n")

# Print summary statistics
cat("=== PRIMER ALIGNMENT SUMMARY ===\n")
cat("Total primers analyzed:", nrow(report), "\n")

cat("\nAlignment count distribution:\n")
alignment_summary <- table(cut(report$num_alignments, 
                              breaks = c(-1, 0, 1, 5, 20, Inf), 
                              labels = c("No hits", "1 hit", "2-5 hits", "6-20 hits", ">20 hits")))
for (bin in names(alignment_summary)) {
  cat("  ", bin, ":", alignment_summary[bin], "\n")
}

# Identify problematic primers
problematic <- report[report$alignment_quality %in% c("POOR", "FAIL"), ]
if (nrow(problematic) > 0) {
  cat("\n=== PROBLEMATIC PRIMERS ===\n")
  cat("Primers with poor/no alignment (consider excluding):\n")
  for (i in 1:min(10, nrow(problematic))) {  # Show max 10
    p <- problematic[i, ]
    cat("  ", p$gene_id, "primer", p$primer_index, "(", p$primer_type, ") -", p$num_alignments, "hits\n")
  }
  if (nrow(problematic) > 10) {
    cat("  ... and", nrow(problematic) - 10, "more (see full report)\n")
  }
}

# Identify best primers
best_primers <- report[report$alignment_quality == "PERFECT", ]
if (nrow(best_primers) > 0) {
  cat("\n=== BEST PRIMERS (PERFECT ALIGNMENT) ===\n")
  cat("Primers with unique transcriptome hits:\n")
  for (i in 1:min(5, nrow(best_primers))) {  # Show max 5
    p <- best_primers[i, ]
    cat("  ", p$gene_id, "primer", p$primer_index, "(", p$primer_type, ") - MAPQ:", p$best_mapq, "\n")
  }
  if (nrow(best_primers) > 5) {
    cat("  ... and", nrow(best_primers) - 5, "more perfect primers\n")
  }
}

File: bin/analyze_primer_alignments.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Analyze primer alignment results from bowtie2 - Python replacement for R script
This script processes BAM files and produces comprehensive alignment analysis
including all metrics from make_alignment_summary.py
"""

import argparse
import pandas as pd
import pysam
from Bio import SeqIO
import sys
import os

def parse_gene_names(transcriptome_fasta=None, transcript_mapping=None):
    """
    Parse gene names from transcriptome FASTA headers or load from mapping file
    Returns dictionary mapping transcript_id -> gene_name
    """
    
    if transcript_mapping and os.path.exists(transcript_mapping):
        print("Loading pre-built transcript-to-gene mapping...")
        try:
            mapping_data = pd.read_csv(transcript_mapping, sep='\t')
            
            if 'transcript_id' not in mapping_data.columns or 'gene_name' not in mapping_data.columns:
                raise ValueError("Mapping file must contain 'transcript_id' and 'gene_name' columns")
            
            transcript_to_gene = dict(zip(mapping_data['transcript_id'], mapping_data['gene_name']))
            print(f"Loaded mapping for {len(transcript_to_gene)} transcripts")
            return transcript_to_gene
            
        except Exception as e:
            print(f"Error loading mapping file: {e}")
            print("Falling back to FASTA parsing...")
    
    # Fallback to FASTA parsing if no mapping file
    if not transcriptome_fasta or not os.path.exists(transcriptome_fasta):
        print("Warning: Transcriptome FASTA not provided or not found. Gene names will not be available.")
        return {}
    
    print("Parsing gene names from transcriptome FASTA (this may take time)...")
    transcript_to_gene = {}
    
    try:
        with open(transcriptome_fasta) as fasta_file:
            for i, record in enumerate(SeqIO.parse(fasta_file, "fasta")):
                # Extract transcript ID (first part before space)
                transcript_id = record.id.split()[0] if ' ' in record.id else record.id
                
                # Extract gene name from header (assuming format like "ENST00000456328 gene=DDX11L2")
                gene_name = "Unknown"
                if "gene=" in record.description:
                    try:
                        gene_part = record.description.split("gene=")[1].split()[0]
                        gene_name = gene_part
                    except IndexError:
                        pass
                
                transcript_to_gene[transcript_id] = gene_name
                
                # Progress indicator for large files
                if (i + 1) % 10000 == 0:
                    print(f"Processed {i + 1} sequences")
        
        print(f"Parsed {len(transcript_to_gene)} transcript-to-gene mappings")
        
        # Debug: show a few example mappings
        test_transcripts = ["ENST00000521262", "ENST00000524349", "ENST00000371437"]
        found_count = 0
        for test_id in test_transcripts:
            if test_id in transcript_to_gene:
                found_count += 1
                if found_count <= 2:
                    print(f"  Example mapping: {test_id} -> {transcript_to_gene[test_id]}")
        
        return transcript_to_gene
        
    except Exception as e:
        print(f"Error parsing transcriptome FASTA: {e}")
        return {}

def analyze_alignments(alignment_bam, primers_tsv, transcriptome_fasta=None, 
                      transcript_mapping=None, out_report="primer_alignment_report.tsv", 
                      out_summary="primer_alignment_summary.tsv"):
    """
    Main function to analyze primer alignments
    """
    
    # Read original primer data
    print("Loading primer data...")
    primers = pd.read_csv(primers_tsv, sep='\t')
    primers['primer_id'] = (primers['gene_id'].astype(str) + "|" +
                           "idx" + primers['primer_index'].astype(str) + "|" +
                           primers['primer_type'].astype(str) + "|" +
                           "strand" + primers['gene_strand'].astype(str))
    
    # Create a lookup dictionary for gene strand information
    primer_to_gene_strand = dict(zip(primers['primer_id'], primers['gene_strand']))
    
    print(f"Loaded {len(primers)} primers")
    
    # Parse transcriptome FASTA for gene name mapping
    transcript_to_gene = parse_gene_names(transcriptome_fasta, transcript_mapping)
    
    # Read alignment results
    print("Reading BAM file...")
    try:
        samfile = pysam.AlignmentFile(alignment_bam, "r")
        
        # Get reference sequence lengths for distance calculations
        ref_lengths = {sq['SN']: sq['LN'] for sq in samfile.header['SQ']}
        
        alignments_data = []
        
        for read in samfile:
            if read.is_unmapped:
                continue
            
            # Only process forward-aligned primers (5' to 3' orientation)
            if read.is_reverse:
                continue
                
            # Extract alignment information including all metrics from make_alignment_summary.py
            gene_id = read.query_name.split("_")[0]
            primer_id_parts = "_".join(read.query_name.split("_")[2:])
            transcript_id = read.reference_name
            transcript_length = ref_lengths.get(transcript_id, None)
            
            # Get gene name from transcript mapping
            gene_name = transcript_to_gene.get(transcript_id, "Unknown")
            
            # Alignment metrics
            alignment_length = read.query_alignment_length if read.query_alignment_length else 0
            start_pos = read.reference_start
            end_pos = read.reference_end if read.reference_end is not None else start_pos + alignment_length
            
            # Distance to end of transcript - strand-specific calculation
            # This represents the total distance from the primer start to the transcript end
            # (includes the primer length itself)
            distance_to_end = "NA"
            if transcript_length is not None:
                # Get gene strand information for this primer
                gene_strand = primer_to_gene_strand.get(read.query_name, "+")  # default to positive
                
                if gene_strand == "-":
                    # For negative strand genes: primers align in reverse orientation
                    # Distance from primer start (leftmost position = 3' end of primer) to transcript start (5' end of gene)
                    # This includes the primer length
                    distance_to_end = start_pos
                else:
                    # For positive strand genes: 
                    # Distance from primer start (leftmost position = 5' end of primer) to transcript end (3' end of gene)
                    # This includes the primer length
                    distance_to_end = transcript_length - start_pos
            
            # Alignment direction - should always be forward after filtering
            direction = "forward"  # Only forward alignments are processed
            
            # Extract mismatches from NM tag
            mismatches = read.get_tag("NM") if read.has_tag("NM") else "NA"
            
            # MAPQ score
            mapq = read.mapping_quality if read.mapping_quality is not None else 255
            
            alignments_data.append({
                'primer_id': read.query_name,
                'gene_id': gene_id,
                'primer_index_parts': primer_id_parts,
                'transcript_id': transcript_id,
                'gene_name': gene_name,
                'gene_strand': primer_to_gene_strand.get(read.query_name, "+"),
                'alignment_start': start_pos,
                'alignment_end': end_pos,
                'alignment_length': alignment_length,
                'mismatches': mismatches,
                'direction': direction,
                'distance_to_end': distance_to_end,
                'mapq': mapq,
                'transcript_length': transcript_length if transcript_length else "NA"
            })
        
        samfile.close()
        
        print(f"Read {len(alignments_data)} alignments from BAM file")
        
    except Exception as e:
        print(f"Error reading BAM file: {e}")
        alignments_data = []
    
    # Convert to DataFrame
    if alignments_data:
        alignment_df = pd.DataFrame(alignments_data)
    else:
        # Create empty DataFrame with correct columns
        alignment_df = pd.DataFrame(columns=['primer_id', 'gene_id', 'primer_index_parts', 
                                           'transcript_id', 'gene_name', 'gene_strand', 'alignment_start', 
                                           'alignment_end', 'alignment_length', 'mismatches', 
                                           'direction', 'distance_to_end', 'mapq', 'transcript_length'])
    
    # Count alignments per primer
    if not alignment_df.empty:
        alignment_counts = alignment_df['primer_id'].value_counts().to_dict()
    else:
        alignment_counts = {}
    
    # Create comprehensive report
    print("Creating alignment report...")
    report = primers.copy()
    report['num_alignments'] = report['primer_id'].map(alignment_counts).fillna(0).astype(int)
    
    # Write report
    report.to_csv(out_report, sep='\t', index=False)
    print(f"Alignment report written to: {out_report}")
    
    # Create detailed alignment summary (all alignments)
    print("Creating detailed alignment summary...")
    
    if not alignment_df.empty:
        # Use all alignments since MAPQ is meaningless with bowtie2 -a
        print(f"Processing {len(alignment_df)} total alignments")
        
        # Merge with primer information
        primer_cols = ['primer_id', 'gene_id', 'primer_index', 'primer_type', 'primer_sequence', 'gene_strand']
        detailed_summary = alignment_df.merge(
            primers[primer_cols], 
            on='primer_id', 
            how='left',
            suffixes=('_align', '_primer')
        )
        
        # Use primer gene_id and gene_strand if alignment versions are missing
        detailed_summary['gene_id'] = detailed_summary['gene_id_primer'].fillna(detailed_summary['gene_id_align'])
        detailed_summary['gene_strand'] = detailed_summary['gene_strand_primer'].fillna(detailed_summary['gene_strand_align'])
        
        # Select and rename columns (remove MAPQ references)
        detailed_summary = detailed_summary.rename(columns={
            'transcript_id': 'aligned_transcript',
            'gene_name': 'aligned_gene_name',
            'direction': 'alignment_strand'
        })
        
        # Reorder columns (remove alignment_mapq)
        output_columns = [
            'gene_id', 'primer_index', 'primer_type', 'primer_sequence', 'gene_strand',
            'aligned_transcript', 'aligned_gene_name', 'alignment_start', 'alignment_end',
            'alignment_length', 'alignment_strand', 'mismatches', 
            'distance_to_end', 'transcript_length'
        ]
        
        # Debug: check which columns are available
        available_columns = detailed_summary.columns.tolist()
        missing_columns = [col for col in output_columns if col not in available_columns]
        if missing_columns:
            print(f"Warning: Missing columns: {missing_columns}")
            print(f"Available columns: {available_columns}")
            # Use only available columns
            output_columns = [col for col in output_columns if col in available_columns]
        
        detailed_summary = detailed_summary[output_columns]
        
        # Sort by gene, primer index, and number of mismatches (best alignments first)
        detailed_summary = detailed_summary.sort_values(['gene_id', 'primer_index', 'mismatches'], 
                                                      ascending=[True, True, True])
        
    else:
        detailed_summary = pd.DataFrame(columns=[
            'gene_id', 'primer_index', 'primer_type', 'primer_sequence', 'gene_strand',
            'aligned_transcript', 'aligned_gene_name', 'alignment_start', 'alignment_end',
            'alignment_length', 'alignment_strand', 'mismatches',
            'distance_to_end', 'transcript_length'
        ])
    
    # Write detailed summary
    detailed_summary.to_csv(out_summary, sep='\t', index=False)
    print(f"Detailed alignment summary written to: {out_summary}")
    
    # Print summary statistics
    print("\n=== PRIMER ALIGNMENT SUMMARY ===")
    print(f"Total primers analyzed: {len(report)}")
    
    print("\nAlignment count distribution:")
    alignment_bins = pd.cut(report['num_alignments'], 
                           bins=[-1, 0, 1, 5, 20, float('inf')], 
                           labels=['No hits', '1 hit', '2-5 hits', '6-20 hits', '>20 hits'])
    alignment_summary = alignment_bins.value_counts()
    for bin_name in alignment_summary.index:
        print(f"  {bin_name}: {alignment_summary[bin_name]}")
    
    # Identify primers with many alignments (potential cross-reactivity)
    many_alignments = report[report['num_alignments'] > 20]
    if not many_alignments.empty:
        print("\n=== PRIMERS WITH MANY ALIGNMENTS ===")
        print("Primers with >20 hits (potential cross-reactivity):")
        for i in range(min(10, len(many_alignments))):
            p = many_alignments.iloc[i]
            print(f"  {p['gene_id']} primer {p['primer_index']} ({p['primer_type']}) - {p['num_alignments']} hits")
        if len(many_alignments) > 10:
            print(f"  ... and {len(many_alignments) - 10} more (see full report)")
    
    # Identify primers with unique alignments (ideal case)
    unique_primers = report[report['num_alignments'] == 1]
    if not unique_primers.empty:
        print("\n=== PRIMERS WITH UNIQUE ALIGNMENTS ===")
        print("Primers with exactly 1 transcriptome hit:")
        for i in range(min(5, len(unique_primers))):
            p = unique_primers.iloc[i]
            print(f"  {p['gene_id']} primer {p['primer_index']} ({p['primer_type']})")
        if len(unique_primers) > 5:
            print(f"  ... and {len(unique_primers) - 5} more unique primers")
    
    print("\nAlignment analysis complete")

def main():
    parser = argparse.ArgumentParser(description="Analyze primer alignment results from bowtie2")
    parser.add_argument('--alignment_bam', required=True, help='Input BAM file from bowtie2 alignment')
    parser.add_argument('--primers_tsv', required=True, help='Original primers TSV file')
    parser.add_argument('--transcriptome_fasta', help='Transcriptome FASTA file for gene name mapping')
    parser.add_argument('--transcript_mapping', help='Pre-built transcript-to-gene mapping TSV file (faster alternative to FASTA)')
    parser.add_argument('--out_report', default='primer_alignment_report.tsv', help='Output alignment report')
    parser.add_argument('--out_summary', default='primer_alignment_summary.tsv', help='Detailed alignment summary per primer')
    
    args = parser.parse_args()
    
    if not args.alignment_bam or not args.primers_tsv:
        print("Error: Both --alignment_bam and --primers_tsv are required")
        sys.exit(1)
    
    analyze_alignments(
        alignment_bam=args.alignment_bam,
        primers_tsv=args.primers_tsv,
        transcriptome_fasta=args.transcriptome_fasta,
        transcript_mapping=args.transcript_mapping,
        out_report=args.out_report,
        out_summary=args.out_summary
    )

if __name__ == "__main__":
    main()
